`Cannot run program "/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/javac"`

When compiling, it cannot find the proper javac path:

```

[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.6.3:testCompile (scala-test-compile-first) on project dp-utilities-mldsl: wrap: java.io.IOException: Cannot run program "/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/javac" (in directory "/home/peng/git/datapassports/dp-utilities"): error=2, No such file or directory -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.6.3:testCompile (scala-test-compile-first) on project dp-utilities-mldsl: wrap: java.io.IOException: Cannot run program "/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/javac" (in directory "/home/peng/git/datapassports/dp-utilities"): error=2, No such file or directory
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:215)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
Caused by: org.apache.maven.plugin.MojoExecutionException: wrap: java.io.IOException: Cannot run program "/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/javac" (in directory "/home/peng/git/datapassports/dp-utilities"): error=2, No such file or directory
    at scala_maven.ScalaMojoSupport.execute (ScalaMojoSupport.java:316)
    at scala_maven.ScalaTestCompileMojo.execute (ScalaTestCompileMojo.java:51)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
Caused by: java.io.IOException: Cannot run program "/usr/lib/jvm/java-8-openjdk-amd64/jre/bin/javac" (in directory "/home/peng/git/datapassports/dp-utilities"): error=2, No such file or directory
    at java.lang.ProcessBuilder.start (ProcessBuilder.java:1048)
    at scala.sys.process.ProcessBuilderImpl$Simple.run (ProcessBuilderImpl.scala:85)
    at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.run (ProcessBuilderImpl.scala:119)
    at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$anonfun$runBuffered$1 (ProcessBuilderImpl.scala:205)
    at scala.runtime.java8.JFunction0$mcI$sp.apply (JFunction0$mcI$sp.scala:17)
    at sbt.internal.inc.javac.JavacLogger.buffer (JavacProcessLogger.scala:49)
    at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.runBuffered (ProcessBuilderImpl.scala:205)
    at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$bang (ProcessBuilderImpl.scala:146)
    at sbt.internal.inc.javac.ForkedJava$.$anonfun$launch$3 (ForkedJava.scala:55)
    at sbt.internal.inc.javac.ForkedJava$.$anonfun$launch$3$adapted (ForkedJava.scala:48)
    at sbt.internal.inc.javac.ForkedJava$.$anonfun$withArgumentFile$1 (ForkedJava.scala:77)
    at sbt.io.IO$.withTemporaryDirectory (IO.scala:490)
    at sbt.io.IO$.withTemporaryDirectory (IO.scala:500)
    at sbt.internal.inc.javac.ForkedJava$.withArgumentFile (ForkedJava.scala:74)
    at sbt.internal.inc.javac.ForkedJava$.launch (ForkedJava.scala:48)
    at sbt.internal.inc.javac.ForkedJavaCompiler.run (ForkedJava.scala:103)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$12 (AnalyzingJavaCompiler.scala:172)
    at scala.runtime.java8.JFunction0$mcV$sp.apply (JFunction0$mcV$sp.scala:18)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler.timed (AnalyzingJavaCompiler.scala:262)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler.compile (AnalyzingJavaCompiler.scala:161)
    at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compileJava$2 (MixedAnalyzingCompiler.scala:103)
    at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compileJava$2$adapted (MixedAnalyzingCompiler.scala:91)
    at sbt.internal.inc.JarUtils$.withPreviousJar (JarUtils.scala:239)
    at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compileJava$1 (MixedAnalyzingCompiler.scala:91)
    at scala.runtime.java8.JFunction0$mcV$sp.apply (JFunction0$mcV$sp.scala:18)
    at sbt.internal.inc.MixedAnalyzingCompiler.timed (MixedAnalyzingCompiler.scala:247)
    at sbt.internal.inc.MixedAnalyzingCompiler.compileJava (MixedAnalyzingCompiler.scala:61)
    at sbt.internal.inc.MixedAnalyzingCompiler.compileJava0$1 (MixedAnalyzingCompiler.scala:197)
    at sbt.internal.inc.MixedAnalyzingCompiler.compile (MixedAnalyzingCompiler.scala:210)
    at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1 (IncrementalCompilerImpl.scala:528)
    at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted (IncrementalCompilerImpl.scala:528)
    at sbt.internal.inc.Incremental$.$anonfun$apply$5 (Incremental.scala:177)
    at sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted (Incremental.scala:175)
    at sbt.internal.inc.Incremental$$anon$2.run (Incremental.scala:461)
    at sbt.internal.inc.IncrementalCommon$CycleState.next (IncrementalCommon.scala:116)
    at sbt.internal.inc.IncrementalCommon$$anon$1.next (IncrementalCommon.scala:56)
    at sbt.internal.inc.IncrementalCommon$$anon$1.next (IncrementalCommon.scala:52)
    at sbt.internal.inc.IncrementalCommon.cycle (IncrementalCommon.scala:263)
    at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8 (Incremental.scala:416)
    at sbt.internal.inc.Incremental$.withClassfileManager (Incremental.scala:503)
    at sbt.internal.inc.Incremental$.incrementalCompile (Incremental.scala:403)
    at sbt.internal.inc.Incremental$.apply (Incremental.scala:169)
    at sbt.internal.inc.IncrementalCompilerImpl.compileInternal (IncrementalCompilerImpl.scala:528)
    at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1 (IncrementalCompilerImpl.scala:482)
    at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError (IncrementalCompilerImpl.scala:332)
    at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally (IncrementalCompilerImpl.scala:420)
    at sbt.internal.inc.IncrementalCompilerImpl.compile (IncrementalCompilerImpl.scala:137)
    at sbt_inc.SbtIncrementalCompiler.compile (SbtIncrementalCompiler.java:186)
    at scala_maven.ScalaCompilerSupport.incrementalCompile (ScalaCompilerSupport.java:364)
    at scala_maven.ScalaCompilerSupport.compile (ScalaCompilerSupport.java:122)
    at scala_maven.ScalaCompilerSupport.doExecute (ScalaCompilerSupport.java:89)
    at scala_maven.ScalaMojoSupport.execute (ScalaMojoSupport.java:310)
    at scala_maven.ScalaTestCompileMojo.execute (ScalaTestCompileMojo.java:51)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
Caused by: java.io.IOException: error=2, No such file or directory
    at java.lang.UNIXProcess.forkAndExec (Native Method)
    at java.lang.UNIXProcess.<init> (UNIXProcess.java:247)
    at java.lang.ProcessImpl.start (ProcessImpl.java:134)
    at java.lang.ProcessBuilder.start (ProcessBuilder.java:1029)
    at scala.sys.process.ProcessBuilderImpl$Simple.run (ProcessBuilderImpl.scala:85)
    at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.run (ProcessBuilderImpl.scala:119)
    at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$anonfun$runBuffered$1 (ProcessBuilderImpl.scala:205)
    at scala.runtime.java8.JFunction0$mcI$sp.apply (JFunction0$mcI$sp.scala:17)
    at sbt.internal.inc.javac.JavacLogger.buffer (JavacProcessLogger.scala:49)
    at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.runBuffered (ProcessBuilderImpl.scala:205)
    at scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$bang (ProcessBuilderImpl.scala:146)
    at sbt.internal.inc.javac.ForkedJava$.$anonfun$launch$3 (ForkedJava.scala:55)
    at sbt.internal.inc.javac.ForkedJava$.$anonfun$launch$3$adapted (ForkedJava.scala:48)
    at sbt.internal.inc.javac.ForkedJava$.$anonfun$withArgumentFile$1 (ForkedJava.scala:77)
    at sbt.io.IO$.withTemporaryDirectory (IO.scala:490)
    at sbt.io.IO$.withTemporaryDirectory (IO.scala:500)
    at sbt.internal.inc.javac.ForkedJava$.withArgumentFile (ForkedJava.scala:74)
    at sbt.internal.inc.javac.ForkedJava$.launch (ForkedJava.scala:48)
    at sbt.internal.inc.javac.ForkedJavaCompiler.run (ForkedJava.scala:103)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$12 (AnalyzingJavaCompiler.scala:172)
    at scala.runtime.java8.JFunction0$mcV$sp.apply (JFunction0$mcV$sp.scala:18)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler.timed (AnalyzingJavaCompiler.scala:262)
    at sbt.internal.inc.javac.AnalyzingJavaCompiler.compile (AnalyzingJavaCompiler.scala:161)
    at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compileJava$2 (MixedAnalyzingCompiler.scala:103)
    at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compileJava$2$adapted (MixedAnalyzingCompiler.scala:91)
    at sbt.internal.inc.JarUtils$.withPreviousJar (JarUtils.scala:239)
    at sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compileJava$1 (MixedAnalyzingCompiler.scala:91)
    at scala.runtime.java8.JFunction0$mcV$sp.apply (JFunction0$mcV$sp.scala:18)
    at sbt.internal.inc.MixedAnalyzingCompiler.timed (MixedAnalyzingCompiler.scala:247)
    at sbt.internal.inc.MixedAnalyzingCompiler.compileJava (MixedAnalyzingCompiler.scala:61)
    at sbt.internal.inc.MixedAnalyzingCompiler.compileJava0$1 (MixedAnalyzingCompiler.scala:197)
    at sbt.internal.inc.MixedAnalyzingCompiler.compile (MixedAnalyzingCompiler.scala:210)
    at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1 (IncrementalCompilerImpl.scala:528)
    at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted (IncrementalCompilerImpl.scala:528)
    at sbt.internal.inc.Incremental$.$anonfun$apply$5 (Incremental.scala:177)
    at sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted (Incremental.scala:175)
    at sbt.internal.inc.Incremental$$anon$2.run (Incremental.scala:461)
    at sbt.internal.inc.IncrementalCommon$CycleState.next (IncrementalCommon.scala:116)
    at sbt.internal.inc.IncrementalCommon$$anon$1.next (IncrementalCommon.scala:56)
    at sbt.internal.inc.IncrementalCommon$$anon$1.next (IncrementalCommon.scala:52)
    at sbt.internal.inc.IncrementalCommon.cycle (IncrementalCommon.scala:263)
    at sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8 (Incremental.scala:416)
    at sbt.internal.inc.Incremental$.withClassfileManager (Incremental.scala:503)
    at sbt.internal.inc.Incremental$.incrementalCompile (Incremental.scala:403)
    at sbt.internal.inc.Incremental$.apply (Incremental.scala:169)
    at sbt.internal.inc.IncrementalCompilerImpl.compileInternal (IncrementalCompilerImpl.scala:528)
    at sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1 (IncrementalCompilerImpl.scala:482)
    at sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError (IncrementalCompilerImpl.scala:332)
    at sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally (IncrementalCompilerImpl.scala:420)
    at sbt.internal.inc.IncrementalCompilerImpl.compile (IncrementalCompilerImpl.scala:137)
    at sbt_inc.SbtIncrementalCompiler.compile (SbtIncrementalCompiler.java:186)
    at scala_maven.ScalaCompilerSupport.incrementalCompile (ScalaCompilerSupport.java:364)
    at scala_maven.ScalaCompilerSupport.compile (ScalaCompilerSupport.java:122)
    at scala_maven.ScalaCompilerSupport.doExecute (ScalaCompilerSupport.java:89)
    at scala_maven.ScalaMojoSupport.execute (ScalaMojoSupport.java:310)
    at scala_maven.ScalaTestCompileMojo.execute (ScalaTestCompileMojo.java:51)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:190)
    at org.apache.maven.lifecycle.internal.builder.multithreaded.MultiThreadedBuilder$1.call (MultiThreadedBuilder.java:186)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.Executors$RunnableAdapter.call (Executors.java:511)
    at java.util.concurrent.FutureTask.run (FutureTask.java:266)
    at java.util.concurrent.ThreadPoolExecutor.runWorker (ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run (ThreadPoolExecutor.java:624)
    at java.lang.Thread.run (Thread.java:748)
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command

```

Here you go. To reproduce the error:
- make sure a java JDK is in the Path, in my case it's `c:\Program Files\Eclipse Adoptium\jdk-8.0.312.7-hotspot\bin\`
- make sure `JAVA_HOME` is not defined
- run `mvn clean compile`
=> following error occurs:
```
[INFO] compiling 1 Scala source and 1 Java source to D:\scala-test-prj\target\classes ...
[WARNING] javac exited with exit code -1
[ERROR] ## Exception when compiling 2 sources to D:\scala-test-prj\target\classes
java.io.IOException: Cannot run program "C:\Program Files\Eclipse Adoptium\jdk-8.0.312.7-hotspot\jre\bin\javac" (in directory "D:\scala-test-prj"): CreateProcess error=2, The system cannot find the file specified
java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
scala.sys.process.ProcessBuilderImpl$Simple.run(ProcessBuilderImpl.scala:85)
scala.sys.process.ProcessBuilderImpl$AbstractBuilder.run(ProcessBuilderImpl.scala:119)
scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$anonfun$runBuffered$1(ProcessBuilderImpl.scala:205)
scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.scala:17)
sbt.internal.inc.javac.JavacLogger.buffer(JavacProcessLogger.scala:49)
scala.sys.process.ProcessBuilderImpl$AbstractBuilder.runBuffered(ProcessBuilderImpl.scala:205)
scala.sys.process.ProcessBuilderImpl$AbstractBuilder.$bang(ProcessBuilderImpl.scala:146)
sbt.internal.inc.javac.ForkedJava$.$anonfun$launch$3(ForkedJava.scala:55)
sbt.internal.inc.javac.ForkedJava$.$anonfun$launch$3$adapted(ForkedJava.scala:48)
sbt.internal.inc.javac.ForkedJava$.$anonfun$withArgumentFile$1(ForkedJava.scala:77)
sbt.io.IO$.withTemporaryDirectory(IO.scala:490)
sbt.io.IO$.withTemporaryDirectory(IO.scala:500)
sbt.internal.inc.javac.ForkedJava$.withArgumentFile(ForkedJava.scala:74)
sbt.internal.inc.javac.ForkedJava$.launch(ForkedJava.scala:48)
sbt.internal.inc.javac.ForkedJavaCompiler.run(ForkedJava.scala:103)
sbt.internal.inc.javac.AnalyzingJavaCompiler.$anonfun$compile$12(AnalyzingJavaCompiler.scala:172)
scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
sbt.internal.inc.javac.AnalyzingJavaCompiler.timed(AnalyzingJavaCompiler.scala:262)
sbt.internal.inc.javac.AnalyzingJavaCompiler.compile(AnalyzingJavaCompiler.scala:161)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compileJava$2(MixedAnalyzingCompiler.scala:103)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compileJava$2$adapted(MixedAnalyzingCompiler.scala:91)
sbt.internal.inc.JarUtils$.withPreviousJar(JarUtils.scala:239)
sbt.internal.inc.MixedAnalyzingCompiler.$anonfun$compileJava$1(MixedAnalyzingCompiler.scala:91)
scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
sbt.internal.inc.MixedAnalyzingCompiler.timed(MixedAnalyzingCompiler.scala:247)
sbt.internal.inc.MixedAnalyzingCompiler.compileJava(MixedAnalyzingCompiler.scala:61)
sbt.internal.inc.MixedAnalyzingCompiler.compileJava0$1(MixedAnalyzingCompiler.scala:197)
sbt.internal.inc.MixedAnalyzingCompiler.compile(MixedAnalyzingCompiler.scala:210)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1(IncrementalCompilerImpl.scala:528)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileInternal$1$adapted(IncrementalCompilerImpl.scala:528)
sbt.internal.inc.Incremental$.$anonfun$apply$5(Incremental.scala:177)
sbt.internal.inc.Incremental$.$anonfun$apply$5$adapted(Incremental.scala:175)
sbt.internal.inc.Incremental$$anon$2.run(Incremental.scala:461)
sbt.internal.inc.IncrementalCommon$CycleState.next(IncrementalCommon.scala:116)
sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:56)
sbt.internal.inc.IncrementalCommon$$anon$1.next(IncrementalCommon.scala:52)
sbt.internal.inc.IncrementalCommon.cycle(IncrementalCommon.scala:263)
sbt.internal.inc.Incremental$.$anonfun$incrementalCompile$8(Incremental.scala:416)
sbt.internal.inc.Incremental$.withClassfileManager(Incremental.scala:503)
sbt.internal.inc.Incremental$.incrementalCompile(Incremental.scala:403)
sbt.internal.inc.Incremental$.apply(Incremental.scala:169)
sbt.internal.inc.IncrementalCompilerImpl.compileInternal(IncrementalCompilerImpl.scala:528)
sbt.internal.inc.IncrementalCompilerImpl.$anonfun$compileIncrementally$1(IncrementalCompilerImpl.scala:482)
sbt.internal.inc.IncrementalCompilerImpl.handleCompilationError(IncrementalCompilerImpl.scala:332)
sbt.internal.inc.IncrementalCompilerImpl.compileIncrementally(IncrementalCompilerImpl.scala:420)
sbt.internal.inc.IncrementalCompilerImpl.compile(IncrementalCompilerImpl.scala:137)
sbt_inc.SbtIncrementalCompiler.compile(SbtIncrementalCompiler.java:186)
scala_maven.ScalaCompilerSupport.incrementalCompile(ScalaCompilerSupport.java:364)
scala_maven.ScalaCompilerSupport.compile(ScalaCompilerSupport.java:122)
scala_maven.ScalaCompilerSupport.doExecute(ScalaCompilerSupport.java:89)
scala_maven.ScalaMojoSupport.execute(ScalaMojoSupport.java:310)
scala_maven.ScalaCompileMojo.execute(ScalaCompileMojo.java:108)
org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:137)
org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:210)
org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:156)
org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:148)
org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:56)
org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:305)
org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:192)
org.apache.maven.DefaultMaven.execute(DefaultMaven.java:105)
org.apache.maven.cli.MavenCli.execute(MavenCli.java:957)
org.apache.maven.cli.MavenCli.doMain(MavenCli.java:289)
org.apache.maven.cli.MavenCli.main(MavenCli.java:193)
sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
java.lang.reflect.Method.invoke(Method.java:498)
org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:282)
org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:225)
org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:406)
org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:347)

[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11.779 s
[INFO] Finished at: 2022-07-11T15:14:37+03:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal net.alchim31.maven:scala-maven-plugin:4.6.3:compile (default-compile) on project scala-test-prj: wrap: java.io.IOException: Cannot run program "C:\Program Files\Eclipse Adoptium\jdk-8.0.312.7-hotspot\jre\bin\javac" (in directory "D:\scala-test-prj"): CreateProcess error=2, The system cannot find the file specified -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

```
[scala-test-prj.zip](https://github.com/davidB/scala-maven-plugin/files/9083855/scala-test-prj.zip)

--- a/src/main/java/util/JavaLocator.java
+++ b/src/main/java/util/JavaLocator.java
@@ -5,6 +5,9 @@
 package util;
 
 import java.io.File;
+import java.io.IOException;
+import java.util.Locale;
+
 import org.apache.maven.toolchain.Toolchain;
 
 /**
@@ -14,29 +17,57 @@ import org.apache.maven.toolchain.Toolchain;
  */
 public class JavaLocator {
 
-  public static String findExecutableFromToolchain(Toolchain toolchain) {
-    String javaExec = null;
+  private static final boolean IS_WINDOWS =
+    System.getProperty("os.name").toLowerCase(Locale.ROOT).contains("win");
+
+  // inspired from org.codehaus.plexus.compiler.javac.JavacCompiler#getJavacExecutable
+  public static String findExecutableFromToolchain(Toolchain toolchain) throws IOException {
 
     if (toolchain != null) {
-      javaExec = toolchain.findTool("java");
+      String fromToolChain = toolchain.findTool("java");
+      if (fromToolChain != null) {
+        return fromToolChain;
+      }
     }
 
-    if (javaExec == null) {
-      String javaHome = System.getenv("JAVA_HOME");
-      if (javaHome == null) {
-        javaHome = System.getProperty("java.home"); // fallback to JRE
+    String javaCommand = "java" + (IS_WINDOWS ? ".exe" : "");
+
+    String javaHomeSystemProperty = System.getProperty("java.home");
+    if (javaHomeSystemProperty != null) {
+      if (javaHomeSystemProperty.endsWith(File.separator + "jre")) {
+        // Old JDK versions contain a JRE. We might be pointing to that.
+        // We want to try to use the JDK instead as we need javac in order to compile mixed Java-Scala projects.
+        File javaExecFile = new File(javaHomeSystemProperty + File.separator + ".." + File.separator + "bin", javaCommand);
+        if (javaExecFile.isFile()) {
+          // getCanonicalPath to get rid of ".."
+          return javaExecFile.getCanonicalPath();
+        }
       }
-      if (javaHome == null) {
-        throw new IllegalStateException(
-            "Couldn't locate java, try setting JAVA_HOME environment variable.");
+
+      // old standalone JRE or modern JDK
+      File javaExecFile = new File(javaHomeSystemProperty + File.separator + "bin", javaCommand);
+      if (javaExecFile.isFile()) {
+        return javaExecFile.getAbsolutePath();
+      } else {
+        throw new IllegalStateException("Couldn't locate java in defined java.home system property.");
       }
-      javaExec = javaHome + File.separator + "bin" + File.separator + "java";
     }
 
-    return javaExec;
+    // fallback: try to resolve from JAVA_HOME
+    String javaHomeEnvVar = System.getenv("JAVA_HOME");
+    if (javaHomeEnvVar == null) {
+      throw new IllegalStateException("Couldn't locate java, try setting JAVA_HOME environment variable.");
+    }
+
+    File javaExecFile = new File(javaHomeEnvVar + File.separator + "bin", javaCommand);
+    if (javaExecFile.isFile()) {
+      return javaExecFile.getAbsolutePath();
+    } else {
+      throw new IllegalStateException("Couldn't locate java in defined JAVA_HOME environment variable.");
+    }
   }
 
-  public static File findHomeFromToolchain(Toolchain toolchain) {
+  public static File findHomeFromToolchain(Toolchain toolchain) throws IOException {
     String executable = findExecutableFromToolchain(toolchain);
     File executableParent = new File(executable).getParentFile();
     if (executableParent == null) {

--- a/src/test/java/util/JavaLocatorTest.java
+++ b/src/test/java/util/JavaLocatorTest.java
@@ -7,7 +7,7 @@ package util;
 import static org.junit.Assert.*;
 
 import java.io.File;
-import java.nio.file.Paths;
+import java.io.IOException;
 import org.apache.maven.toolchain.Toolchain;
 import org.junit.Rule;
 import org.junit.Test;
@@ -18,23 +18,20 @@ public class JavaLocatorTest {
   @Rule public final EnvironmentVariables environmentVariables = new EnvironmentVariables();
 
   @Test
-  public void shouldReturnNullWhenJavaIsNotAvailableOnCommandLineAndJavaHomeIsPresent() {
-    Toolchain toolchain = new NullReturningToolChain();
-    environmentVariables.set("JAVA_HOME", "test");
-    assertEquals(
-        Paths.get("test", "bin", "java").toString(),
-        JavaLocator.findExecutableFromToolchain(toolchain));
+  public void shouldReturnNotNullWhenJavaIsNotAvailableOnCommandLineAndJavaHomeIsPresent() throws IOException {
+    Toolchain toolchain = new ReturningToolChain(null);
+    assertNotNull(JavaLocator.findExecutableFromToolchain(toolchain));
   }
 
   @Test
   public void shouldReturnPathToJavaWhenJavaIsPresent() throws Exception {
-    Toolchain toolchain = new ReturningToolChain();
+    Toolchain toolchain = new ReturningToolChain("my-path-to-java");
     assertEquals("my-path-to-java", JavaLocator.findExecutableFromToolchain(toolchain));
   }
 
   @Test
-  public void shouldThrowExceptionWhenNothingCouldBeFound() {
-    Toolchain toolchain = new NullReturningToolChain();
+  public void shouldThrowExceptionWhenNothingCouldBeFound() throws IOException {
+    Toolchain toolchain = new ReturningToolChain(null);
     System.clearProperty("java.home");
     environmentVariables.set("JAVA_HOME", null);
     try {
@@ -47,18 +44,23 @@ public class JavaLocatorTest {
   }
 
   @Test
-  public void shouldReturnParentOfChildOfJavaHomeFolder() {
-    File home = JavaLocator.findHomeFromToolchain(new TestStringReturningToolChain());
+  public void shouldReturnParentOfChildOfJavaHomeFolder() throws IOException {
+    File home = JavaLocator.findHomeFromToolchain(new ReturningToolChain("parent/child/my-path-to-java"));
     assertEquals("parent", home.getPath());
   }
 
   @Test
-  public void shouldReturnNullWhenFileIsNotPresent() {
-    File home = JavaLocator.findHomeFromToolchain(new ReturningToolChain());
+  public void shouldReturnNullWhenFileIsNotPresent() throws IOException {
+    File home = JavaLocator.findHomeFromToolchain(new ReturningToolChain("my-path-to-java"));
     assertNull(home);
   }
 
-  class NullReturningToolChain implements Toolchain {
+  static final class ReturningToolChain implements Toolchain {
+    private final String tool;
+
+    ReturningToolChain(String tool) {
+      this.tool = tool;
+    }
 
     @Override
     public String getType() {
@@ -66,34 +68,8 @@ public class JavaLocatorTest {
     }
 
     @Override
-    public String findTool(String s) {
-      return null;
-    }
-  }
-
-  class TestStringReturningToolChain implements Toolchain {
-
-    @Override
-    public String getType() {
-      return null;
-    }
-
-    @Override
-    public String findTool(String s) {
-      return "parent/child/my-path-to-java";
-    }
-  }
-
-  class ReturningToolChain implements Toolchain {
-
-    @Override
-    public String getType() {
-      return null;
-    }
-
-    @Override
-    public String findTool(String s) {
-      return "my-path-to-java";
+    public String findTool(String toolName) {
+      return tool;
     }
   }
 }

